### PR TITLE
CSHARP-4153: Add FLE 2 API to AutoEncryptionOpts.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

**Review by BorisDog — APPROVED** [review]
LGTM. Notice the minor comment.

File: bindings/cs/MongoDB.Libmongocrypt/CryptOptions.cs [hunk 1]
```diff
@@ -41,6 +43,18 @@ public CryptOptions(IEnumerable<KmsCredentials> credentials) : this(credentials,
             Schema = schema;
         }
 
+        public CryptOptions(
```
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/CryptOptions.cs:46]: done

File: bindings/cs/MongoDB.Libmongocrypt/PinnedBinary.cs [hunk 2]
```diff
@@ -24,6 +24,23 @@ namespace MongoDB.Libmongocrypt
     /// <seealso cref="Binary" />
     internal class PinnedBinary : Binary
     {
+        #region static
+        internal static void HandleAsPinnedBinary<THandle>(THandle handle, byte[] bytes, Status status, Func<THandle, BinarySafeHandle, bool> handleFunc) where THandle : CheckableSafeHandle
```
Comment (BorisDog) [inline bindings/cs/MongoDB.Libmongocrypt/PinnedBinary.cs:28]: minor: `Handle` naming is a bit unclear. But I don't have better suggestions, maybe `HandleCheck`?

File: bindings/cs/MongoDB.Libmongocrypt/CryptClient.cs [hunk 3]
```diff
@@ -136,33 +123,55 @@ public CryptContext StartExplicitEncryptionContextWithKeyId(byte[] keyId, string
...
-                }
+                PinnedBinary.HandleAsPinnedBinary(handle, keyId, _status, (h, pb) => Library.mongocrypt_ctx_setopt_key_id(h, pb));
+            }
+            else if (keyAltName != null)
+            {
+                PinnedBinary.HandleAsPinnedBinary(handle, keyAltName, _status, (h, pb) => Library.mongocrypt_ctx_setopt_key_alt_name(h, pb));
             }
 
-            handle.Check(_status, Library.mongocrypt_ctx_setopt_algorithm(handle, encryptionAlgorithm, -1));
+            switch (encryptionAlgorithm)
```
Comment (BorisDog) [inline bindings/cs/MongoDB.Libmongocrypt/CryptClient.cs:149]: Does this need to be case-sensitive, and why?

Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClient.cs:149]: good question, initially I was going to agree with you and change it to support case insensitive check here, but then I noticed that handling old algorithm names are case sensitive on c level https://github.com/mongodb/libmongocrypt/blob/master/src/mongocrypt-ctx.c#L292. So I think we should follow this logic for new algorithms as well. Making it case insensitive should happen together with changing a libmongocrypt level as well which I think makes sense but it should be a diff ticket

Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClient.cs:149]: NOTE: spec says says about case sensitive check as well

File: bindings/cs/MongoDB.Libmongocrypt/PinnedBinary.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/PinnedBinary.cs:28]: The right name here probably should be like: `RunWithBytesAsPinnedBinary`, maybe we can simplify it to: `RunAsPinnedBinary`?

**Pull request merged**